src/Sylius/Bundle/UiBundle/Renderer/TemplateEventRenderer.php

```diff
@@ -0,0 +1,43 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\UiBundle\Renderer;
+
+use Sylius\Bundle\UiBundle\Registry\TemplateBlockRegistryInterface;
+
+final class TemplateEventRenderer implements TemplateEventRendererInterface
+{
+    /** @var TemplateBlockRegistryInterface */
+    private $templateBlockRegistry;
+
+    /** @var TemplateBlockRendererInterface */
+    private $templateBlockRenderer;
+
+    public function __construct(TemplateBlockRegistryInterface $templateBlockRegistry, TemplateBlockRendererInterface $templateBlockRenderer)
+    {
+        $this->templateBlockRegistry = $templateBlockRegistry;
+        $this->templateBlockRenderer = $templateBlockRenderer;
+    }
+
+    public function render(string $eventName, array $context = []): string
+    {
+        $templateBlocks = $this->templateBlockRegistry->findEnabledForEvent($eventName);
+        $renderedTemplates = [];
+
+        foreach ($templateBlocks as $templateBlock) {
+            $renderedTemplates[] = $this->templateBlockRenderer->render($eventName, $templateBlock, $context);
```

Comment:
Why not concatenate a string instead. My guess is you'd want a trailing new line anyways in the final source code.

So like:
`$str .= $this->templateBlockRenderer->render($eventName, $templateBlock, $context) . "\n";`

Reply:
Concat in this case could be faster https://stackoverflow.com/questions/4502654/php-many-concats-or-one-implode

Reply:
I think we don't need trailing new line:

```twig
<div>
  {{ sylius_template_event('event') }}
</div>
```

With current implementation it will return:
```
<div>
  <p>First</p>
  <p>Second</p>
</div>
```

With trailing new line:
```
<div>
  <p>First</p>
  <p>Second</p>

</div>
```

Reply:
As for the performance concerns, I think using implode is clearer than concatenation and right trimming last newline character - those comments are 9 years old, based on PHP 5 and still seem to be rather a microoptimisation.